c-dependencies/js-compute-runtime/builtins/backend.cpp

```diff
@@ -17,6 +19,617 @@
+
+    /* Add everything else */
+    result.insert(result.end(), ecdh.begin(), ecdh.end());
+    result.insert(result.end(), ciphers.begin(), ciphers.end());
+
+    /* Move ciphers without forward secrecy to the end */
+    auto ciphersWithoutForwardSecrecy = filterByKeyExchange(result, std::set{KeyExchange::RSA});
+    moveToEnd(result, ciphersWithoutForwardSecrecy);
+
+    strengthSort(result);
+    return result;
+  }
+
+  std::vector<Cipher> filterByStrengthBits(std::vector<Cipher> &ciphers, int strength_bits) {
+    std::vector<Cipher> result;
+    for (auto cipher : ciphers) {
+      if (cipher.getStrength_bits() == strength_bits) {
+        result.push_back(cipher);
+      }
+    }
+    return result;
+  }
+
+  std::vector<Cipher> filterByProtocol(std::vector<Cipher> &ciphers, std::set<Protocol> protocol) {
+    return filter(ciphers, protocol, std::nullopt, std::nullopt, std::nullopt, std::nullopt,
+                  std::nullopt);
+  }
+
+  std::vector<Cipher> filterByKeyExchange(std::vector<Cipher> &ciphers, std::set<KeyExchange> kx) {
+    return filter(ciphers, std::nullopt, kx, std::nullopt, std::nullopt, std::nullopt,
+                  std::nullopt);
+  }
+
+  std::vector<Cipher> filterByAuthentication(std::vector<Cipher> &ciphers,
+                                             std::set<Authentication> au) {
+    return filter(ciphers, std::nullopt, std::nullopt, au, std::nullopt, std::nullopt,
+                  std::nullopt);
+  }
+
+  std::vector<Cipher> filterByEncryption(std::vector<Cipher> &ciphers, std::set<Encryption> enc) {
+    return filter(ciphers, std::nullopt, std::nullopt, std::nullopt, enc, std::nullopt,
+                  std::nullopt);
+  }
+
+  std::vector<Cipher> filterByEncryptionLevel(std::vector<Cipher> &ciphers,
+                                              std::set<EncryptionLevel> level) {
+    return filter(ciphers, std::nullopt, std::nullopt, std::nullopt, std::nullopt, level,
+                  std::nullopt);
+  }
+
+  std::vector<Cipher> filterByMessageDigest(std::vector<Cipher> &ciphers,
+                                            std::set<MessageDigest> mac) {
+    return filter(ciphers, std::nullopt, std::nullopt, std::nullopt, std::nullopt, std::nullopt,
+                  mac);
+  }
+
+  std::vector<Cipher>
+  filter(std::vector<Cipher> &ciphers, std::optional<std::set<Protocol>> protocols,
+         std::optional<std::set<KeyExchange>> kx, std::optional<std::set<Authentication>> au,
+         std::optional<std::set<Encryption>> enc, std::optional<std::set<EncryptionLevel>> level,
+         std::optional<std::set<MessageDigest>> mac) {
+    std::vector<Cipher> result;
+    for (auto cipher : ciphers) {
+      if (protocols.has_value() && protocols->find(cipher.getProtocol()) != protocols->end()) {
+        result.push_back(cipher);
+      }
+      if (kx.has_value() && kx->find(cipher.getKx()) != kx->end()) {
+        result.push_back(cipher);
+      }
```

Comment:
Won't this potentially duplicate entries in the result vector? What do you think about ending each conditional with `continue`?